ouroboros-consensus-cardano/test/Test/ThreadNet/Cardano.hs

```diff
@@ -530,20 +577,57 @@ prop_simple_cardano_convergence TestSetup
         w :: Word64
         w = Shelley.computeStabilityWindow setupK coeff
 
-        -- the independent probability that a Shelley slot will have at least
-        -- one leader
-        f :: Double
-        f = fromRational $ SL.unitIntervalToRational $ SL.activeSlotVal coeff
-
-        -- a strong threshold, along the lines of " not even once in the
-        -- lifetime of the project "
-        oneBillionth :: Double
-        oneBillionth = 1e-9
+    -- Whether there was a Shelley block forged in a non-overlay slot.
+    --
+    -- This event evidences that the stake pools were correctly created and
+    -- delegated to.
+    label_hadActiveNonOverlaySlots :: String
+    label_hadActiveNonOverlaySlots =
+        show $ or $
+        [ Set.notMember slot overlaySlots
+        | (_nid, no) <- Map.toList testOutputNodes
+        , let NodeOutput{nodeOutputForges} = no
+        , (slot, blk) <- Map.toDescList nodeOutputForges
+        , isShelley blk
+        ]
+      where
+        TestOutput{testOutputNodes} = testOutput
 
-        isShelley :: CardanoBlock c -> Bool
-        isShelley = \case
-            BlockByron{}   -> False
-            BlockShelley{} -> True
+    -- All OBFT overlay slots in the test.
+    overlaySlots :: Set SlotNo
+    overlaySlots =
+        Set.filter (< SlotNo t) $
+        Set.unions $
+        takeWhile (isJust . Set.lookupLT (SlotNo t)) $
+        iterate (shift epochSizeShelley) $   -- NB iterate begins with id
+        shift epochSizeByron $
+        overlayOffsets
+      where
+        NumSlots t = numSlots
+
+        shift :: EpochSize -> Set SlotNo -> Set SlotNo
+        shift x = Set.mapMonotonic (Util.addSlots (unEpochSize x))
+
+        -- The relative offset of the overlay slots in each Shelley epoch.
+        overlayOffsets :: Set SlotNo
+        overlayOffsets =
+            Map.keysSet $
+            runIdentity $ flip runReaderT shelleyGlobals $
+            SL.overlaySchedule
+              @Crypto
+              (EpochNo 0)   -- Ignoring this input; see uses of @shift@ above
```

Comment:
I don't understand, if you're shifting all slots by the epoch size, why not use `EpochNo 1` here?

Ah, you're shifting by `epochSizeByron` first. A cleaner solution would be to pass the correct `EpochInfo` to `shelleyGlobals`. You would have to construct a correct `Summary` and pass that to [`snapshotEpochInfo`](https://github.com/input-output-hk/ouroboros-network/blob/3a83c64ce76fb995b9557d3eb55d3c7e1c2fc251/ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/EpochInfo.hs#L48).

Reply:
We agreed that the `Summary` is not easy to create here, so to avoid inflating the diff eta, we compromised with a smaller refactor:

```
        -- The overlay slots in the ith Shelley epoch.
        overlayOffsets :: Word64 -> Set SlotNo
        overlayOffsets i =
            -- Shift to account for the initial Byron era.
            Set.mapMonotonic (Util.addSlots numByronSlots) $

            Map.keysSet $
            runIdentity $ flip runReaderT shelleyGlobals $
            SL.overlaySchedule
              @Crypto
              (EpochNo i)   -- NB 0 <-> the first /Shelley/ epoch
              genesisKeyHashes
              (sgProtocolParams genesisShelley)   -- notably contains setupD
```